### ISSUE TITLE
Defer loading of javascript resources instead moving to bottom

### ISSUE THREAD

**Issue by djmj** [primefaces/primefaces#10679]
### Description

Isn't "moving scripts to bottom" right before _</body>_ closing tag a workaround before _defer_ attribute was introduced for script resource loading?

I read a lot lately about async, defer and move to bottom and using the defer attribute is similar then moving to bottom but has advantage that the browser can download the resource files in parallel while the DOM is parsed, especially if a different connection is used like resources are on a CDN.

Moving scripts to bottom all the DOM needs to be parsed completely until the browser finally knows the script tag and after this the downloading of the files starts.

**Research**

- https://web.dev/efficiently-load-third-party-javascript/
- https://javascript.plainenglish.io/loading-scripts-with-async-and-defer-dfc9dc46077b

### Describe the solution you would like

Instead of moving scripts to bottom, just add defer attribute and keep them in the head.

**Comment by melloware** [comment]
OmniFaces already has a DeferredScript: https://showcase.omnifaces.org/components/deferredScript

**Comment by djmj** [comment]
Comparison is only fair between current _MoveScriptsToBottom_ and _defer_ attribute. This issue would provide a primefaces only solution instead and also simplify the move scripts to bottom code.

**DeferredScript**

Yeah i also read about this but not yet sure if i want to use it and i see this issue more about a primefaces only improvement. I dont want to use _DeferredScript_ cause noone provided the _HeadRenderer_ which may be prone to future PrimeFaces changes and we must maintain a js file and load another one.

I will consider _DeferredScript_ after testing _defer_ attribute. 

**Custom ResponseWriterWrapper**

Can it be solved with a custom _ResponseWriterWrapper_ that will simply add _defer_ to all javacsript resources. Not sure since i have never done it. Just seeing https://github.com/primefaces/primefaces/blob/64eb57b58dd2b19e0eeca6c8caca6b82bf253e57/primefaces/src/main/java/org/primefaces/context/PrimeFacesContext.java#L90

**Comment by melloware** [comment]
You could try and model a repsonse writer after move scripts to bottom but instead just add the defer="true" to see if it works.

**Comment by tandraschko** [comment]
MovingToBottom has 2 features:
1) moving scripts includes to the bottom
2) merge inline scripts and minify

it sounds like 1) can be improved with deffered and leaving it in the head, not sure if it works combined with 2).

**Comment by melloware** [comment]
We should also figure out what can be deferred and what should not be by default?  I know deferring can be tricky because of 

> "Usually, it are those kind of scripts which are just for progressive enhancement and thus not essential for the functioning of the webpage."

So we have to be careful that not everything can be deferred?

**Comment by djmj** [comment]
@melloware Yeah i guess so. It seems it could happen that jquery was not yet loaded before the inline scripts are executed.

Some say its possible to defer them aswell with _<script type="modul">_ . https://stackoverflow.com/questions/41394983/how-to-defer-inline-javascript

This needs some testing and here is some enhanced background:

https://github.com/WordPress/performance/issues/102

**Comment by melloware** [comment]
I think if PF is going to do this we will probably at least know which scripts can and can't be deferred. Like Jquery can't be deferred but then everything else should be assumed to be deferred and we can keep testing until we get it right.

**Comment by djmj** [comment]
That sounds great. Are you sure about jquery being not deferable? If the inline scripts are moved at last position and also deferred i think it should work cause defer scripts are guaranteed to execute in the order that they appear in the document.

**Comment by melloware** [comment]
@djmj I am basing my info on this amazing BalusC post where he talks about deferring PrimeFaces JS.  This post is quite old but it does mention you can't defer jQuery: https://stackoverflow.com/a/23173640/502366

**Comment by melloware** [comment]
OK I submitted a PR based on this post: https://stackoverflow.com/questions/12416307/is-it-necessary-to-put-scripts-at-the-bottom-of-a-page-when-using-the-defer-at

Which says you should move to bottom AND defer.

**Comment by djmj** [comment]
I read the answers differently since two of the top voted answers of this question like https://stackoverflow.com/a/12416377 say:

Keep it in head and add defer attribute to benefit from browser seeing the script src sooner and download it in parallel.


For own testing purpose i will also test this myself in future with our online shops but i will do it maybe late next year or so since our speed improvement goal (get green at google core vitals) is already reached with help of moving the scripts to bottom so we will focus now on other things.

**Comment by melloware** [comment]
Yep so some say the way you are talking about is the optimal way but us moving to bottom with deferred will still be negligee compared to the first way. So the easiest is to added deferred to what we are doing.

**Comment by djmj** [comment]
Yeah and thats why i shifted focus already cause even without the "optimal" way we still get great enough results now in our speed tests.
